libpermian/plugins/anaconda_webui/__init__.py

```diff
@@ -0,0 +1,498 @@
+    """
+
+    instances = {}
+    temp_dirs = {}
+    npm_dependencies = ('chrome-remote-interface', 'sizzle')
+
+    @classmethod
+    def factory(cls, testRuns, crcList):
+        """
+        Make separate instances of this workflow for given crcIds.
+
+        Also create Semaphore that will limit the number of VMs running at once
+        and threading lock used for setup and teardown
+
+        :param crcIds: List of CaseRunConfiguration which belong to this workflow.
+        :type crcIds: list
+        """
+        vm_limit = testRuns.settings.getint('AnacondaWebUI', 'hypervisor_vm_limit')
+        vm_semaphore = threading.Semaphore(vm_limit)
+        setup_lock = threading.Lock()
+
+        for singleCrcList in crcList.by_key(lambda x: x.id).values():
+            cls(testRuns, singleCrcList, setup_lock, vm_semaphore)
+
+    def __init__(self, testRuns, crcList, setup_lock, vm_semaphore):
+        super().__init__(testRuns, crcList)
+        self.installation_source = self.event.InstallationSource
+        self.vm_semaphore = vm_semaphore
+        self.setup_lock = setup_lock
+        self.vm_name = f'anaconda-webui-{hash(self.crc)}'
+        self.vm_ip = None
+        self.last_log = ''
+        self.use_container = self.settings.getboolean('AnacondaWebUI', 'use_container')
+        self.architecture = self.crc.configuration.get('architecture')
+
+        self.git_anaconda_repo = self.settings.get('AnacondaWebUI', 'anaconda_repo')
+        self.git_anaconda_branch = self.crc.configuration['branch']
+        self.git_cockpit_repo = self.settings.get('AnacondaWebUI', 'cockpit_repo')
+        self.git_cockpit_branch = self.settings.get('AnacondaWebUI', 'cockpit_branch')
+        self.git_bots_repo = self.settings.get('AnacondaWebUI', 'bots_repo')
+        self.git_bots_branch = self.settings.get('AnacondaWebUI', 'bots_branch')
+
+        self.port_ssh = self.settings.get('AnacondaWebUI', 'port_ssh')
+        self.port_webui = self.settings.get('AnacondaWebUI', 'port_webui')
+        self.webui_location = self.settings.get('AnacondaWebUI', 'webui_location')
+
+        automation_data = self.crc.testcase.execution.automation_data
+
+        self.test_file = automation_data['script_file']
+        self.test_case_name = automation_data['test_case']
+
+        self.test_repo_dir = None
+        self.test_repo_name = automation_data.get('test_repo')
+        if self.test_repo_name:
+            self.test_repo_url = self.settings.get('AnacondaWebUIRepos', self.test_repo_name)
+
+        self.additional_repos = automation_data.get('additional_repos')
+
+        self.kernel_cmdline = automation_data.get('kernel_cmdline')
+
+        self.webui_startup_timeout = int(automation_data.get('webui_startup_timeout', 20))
+
+        self.hypervisor_host = self.settings.get('VMHypervisors', self.architecture)
+
+    def setup(self):
+        # Check if we have hypervisor for specified architecture
+        if self.hypervisor_host is None:
```

Comment:
In current situation, this condition is always false. If the section doesn't exist, the `self.settings.get('VMHypervisors', self.architecture)` raises exception => the setup cannot be executed. If there's no value assigned, the resulting value is empty string which is not `None`.